File: src/api/services/daemon.py
import threading
import time
import os
import logging
from datetime import datetime, timedelta
from sqlalchemy import func
from sqlalchemy.orm import aliased
from ..database import SessionLocal
from ..models import Member, CheckResult
from ..era_models import ERAUpload, ERAMatch
from ..dependencies import browser_pool
from .era import sync_era_files, verify_and_repair_era_file, ERA_UPLOAD_DIR

logger = logging.getLogger(__name__)

auto_check_daemon_running = False
auto_check_thread = None

# Track last verification time (persists across daemon iterations)
_last_era_verification = None

def auto_check_daemon():
    """Background daemon that automatically queues unchecked, failed, and captcha members."""
    global auto_check_daemon_running, _last_era_verification
    logger.info("Auto-check daemon started")
    
    retry_interval_minutes = 60
    verification_interval_hours = 24  # Run ERA verification every 24 hours
    last_retry_check = time.time()
    
    first_run = True
    
    while auto_check_daemon_running:
        try:
            db = SessionLocal()
            current_time = time.time()
            
            # 0. Sync ERA files (Auto-parsing)
            try:
                # Daemon thread spawns parsing threads if needed
                # On first run, force resume of any 'parsing' files (zombies from restart)
                sync_era_files(db, background_tasks=None, startup_mode=first_run)
                first_run = False
            except Exception as e:
                logger.error(f"ERA sync error: {e}")
            
            # 0.5 Periodic ERA Verification (self-healing check every 24 hours)
            try:
                now = datetime.utcnow()
                should_verify = (
                    _last_era_verification is None or 
                    (now - _last_era_verification) > timedelta(hours=verification_interval_hours)
                )
                
                if should_verify:
                    # Only verify 'complete' uploads to avoid interfering with active parsing
                    complete_uploads = db.query(ERAUpload).filter(ERAUpload.status == 'complete').all()
                    
                    if complete_uploads:
                        logger.info(f"Periodic ERA verification: Checking {len(complete_uploads)} complete uploads...")
                        
                        for upload in complete_uploads:
                            file_path = os.path.join(ERA_UPLOAD_DIR, upload.filename)
                            if os.path.exists(file_path):
                                repaired = verify_and_repair_era_file(db, file_path, upload)
                                if repaired > 0:
                                    logger.warning(f"Periodic verification repaired {repaired} records for {upload.filename}")
                        
                        _last_era_verification = now
                        logger.info("Periodic ERA verification complete")
                    else:
                        # No complete uploads, but still mark as checked
                        _last_era_verification = now
                        
            except Exception as e:
                logger.error(f"Periodic ERA verification error: {e}")
            
            # 1. Find all members that have never been checked
            # Use explicit select() to avoid SAWarning about coercing subquery
            stmt = db.query(CheckResult.member_id).distinct()
            unchecked_members = db.query(Member).filter(
                ~Member.id.in_(stmt)
            ).all()
            
            if unchecked_members:
                logger.info(f"Auto-check daemon found {len(unchecked_members)} unchecked members")
                for member in unchecked_members:
                    try:
                        browser_pool.enqueue_check(member.id)
                        logger.debug(f"Auto-queued unchecked member {member.id}")
                    except Exception as e:
                        logger.error(f"Failed to auto-queue member {member.id}: {e}")
            
            # 2. Continuously retry failed, partial, and captcha members
            # We rely on browser_pool.queued_items to prevent duplicate queuing
            logger.debug("Auto-check daemon checking for failed/partial/captcha members to retry")
            
            # Subquery to get the latest check result for each member (using MAX(id) for consistency)
            latest_checks_subquery = db.query(
                CheckResult.member_id,
                func.max(CheckResult.id).label("max_id")
            ).group_by(CheckResult.member_id).subquery()
            
            LatestCheckResult = aliased(CheckResult)
            
            # Retry only if the result is older than 60 minutes to prevent infinite loops
            retry_threshold = datetime.utcnow() - timedelta(minutes=retry_interval_minutes)
            
            # Get members whose latest check was Captcha, Fail, or missing ERA match despite being verified
            retry_results = db.query(LatestCheckResult).join(
                latest_checks_subquery,
                LatestCheckResult.id == latest_checks_subquery.c.max_id
            ).filter(
                (LatestCheckResult.timestamp < retry_threshold) &
                (
                    (LatestCheckResult.result.in_(['Captcha', 'Fail', 'Fail_Suburb', 'Fail_Street', 'Fail_No_Match'])) |
                    (LatestCheckResult.result.in_(['Pass', 'Partial']) & ~db.query(ERAMatch).filter(ERAMatch.member_id == LatestCheckResult.member_id).exists())
                )
            ).all()
            
            if retry_results:
                logger.info(f"Auto-check daemon found {len(retry_results)} members to retry")
                for result in retry_results:
                    try:
                        member = db.query(Member).filter(Member.id == result.member_id).first()
                        if member:
                            browser_pool.enqueue_check(member.id)
                    except Exception as e:
                        logger.error(f"Failed to auto-queue retry for member {result.member_id}: {e}")
            
            db.close()
            time.sleep(30)
            
        except Exception as e:
            logger.error(f"Auto-check daemon error: {e}")
            time.sleep(60)
    
    logger.info("Auto-check daemon stopped")

def start_daemon():
    global auto_check_daemon_running, auto_check_thread
    if not auto_check_daemon_running:
        auto_check_daemon_running = True
        auto_check_thread = threading.Thread(target=auto_check_daemon, daemon=True)
        auto_check_thread.start()
        logger.info("Auto-check daemon thread started")

def stop_daemon():
    global auto_check_daemon_running
    auto_check_daemon_running = False
    if auto_check_thread:
        auto_check_thread.join(timeout=5)

File: rust/src/api/daemon.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use sqlx::SqlitePool;
use tokio::sync::Mutex;
use tracing::{error, info};

use crate::api::worker_pool::BrowserPool;

/// Interval between unchecked-member sweeps.
const SWEEP_INTERVAL: Duration = Duration::from_secs(30);

/// Interval between retries of failed/captcha members.
const RETRY_INTERVAL: Duration = Duration::from_secs(3600); // 1 hour

/// Background daemon that automatically queues unchecked members for AEC verification
/// and periodically retries failed or captcha-blocked members.
pub struct AutoCheckDaemon {
    pool: SqlitePool,
    browser_pool: Arc<BrowserPool>,
    running: Arc<Mutex<bool>>,
}

impl AutoCheckDaemon {
    pub fn new(pool: SqlitePool, browser_pool: Arc<BrowserPool>) -> Self {
        AutoCheckDaemon {
            pool,
            browser_pool,
            running: Arc::new(Mutex::new(false)),
        }
    }

    /// Start the daemon in a background tokio task.
    pub fn start(self: Arc<Self>) {
        let daemon = Arc::clone(&self);
        tokio::spawn(async move {
            daemon.run().await;
        });
    }

    /// Signal the daemon to stop.
    pub async fn stop(&self) {
        let mut running = self.running.lock().await;
        *running = false;
        info!("AutoCheckDaemon stop signal sent.");
    }

    async fn run(&self) {
        {
            let mut running = self.running.lock().await;
            *running = true;
        }
        info!("AutoCheckDaemon started");

        let mut last_retry = Instant::now();

        loop {
            // Check if we should keep running
            {
                let running = self.running.lock().await;
                if !*running {
                    break;
                }
            }

            // 1. Find and queue unchecked members
            if let Err(e) = self.queue_unchecked_members().await {
                error!("Daemon error queuing unchecked members: {}", e);
            }

            // 2. Periodically retry failed/captcha/incomplete members
            if last_retry.elapsed() >= RETRY_INTERVAL {
                if let Err(e) = self.queue_retry_members().await {
                    error!("Daemon error queuing retry members: {}", e);
                }
                last_retry = Instant::now();
            }

            tokio::time::sleep(SWEEP_INTERVAL).await;
        }

        info!("AutoCheckDaemon stopped.");
    }

    /// Queue all members that have never had a check result.
    async fn queue_unchecked_members(&self) -> anyhow::Result<()> {
        let unchecked: Vec<(i64,)> = sqlx::query_as(
            "SELECT id FROM members WHERE id NOT IN (SELECT DISTINCT member_id FROM check_results)",
        )
        .fetch_all(&self.pool)
        .await?;

        if !unchecked.is_empty() {
            info!("Daemon found {} unchecked member(s)", unchecked.len());
            for (member_id,) in unchecked {
                if let Err(e) = self.browser_pool.enqueue_check(member_id) {
                    error!("Daemon failed to enqueue member {}: {}", member_id, e);
                }
            }
        }

        Ok(())
    }

    /// Queue members whose latest check result is Captcha, Fail variant,
    /// or a Pass with missing electorate information.
    async fn queue_retry_members(&self) -> anyhow::Result<()> {
        info!("Daemon performing periodic retry sweep...");

        // Get latest check result per member using a subquery
        let retry_rows: Vec<(i64,)> = sqlx::query_as(
            r#"
            SELECT cr.member_id
            FROM check_results cr
            INNER JOIN (
                SELECT member_id, MAX(timestamp) AS max_ts
                FROM check_results
                GROUP BY member_id
            ) latest ON cr.member_id = latest.member_id AND cr.timestamp = latest.max_ts
            WHERE cr.result IN ('Captcha', 'Fail', 'Fail_Suburb', 'Fail_Street', 'Fail_No_Match')
               OR (cr.result = 'Pass' AND (cr.federal_division IS NULL OR cr.federal_division = ''))
            "#,
        )
        .fetch_all(&self.pool)
        .await?;

        if !retry_rows.is_empty() {
            info!("Daemon retrying {} failed/captcha/incomplete member(s)", retry_rows.len());
            for (member_id,) in retry_rows {
                if let Err(e) = self.browser_pool.enqueue_check(member_id) {
                    error!("Daemon failed to enqueue retry for member {}: {}", member_id, e);
                }
            }
        }

        Ok(())
    }
}
